truffle/com.oracle.truffle.api/src/com/oracle/truffle/api/nodes/Node.java

```diff
@@ -517,6 +517,17 @@ public String getLanguage() {
         return "";
     }
 
+    /**
+     * Reports the execution count of a loop that is a child of this node. The optimization
+     * heuristics can use the loop count to guide compilation and inlining.
+     *
+     * @count count the number iterations to report to the runtime system
+     */
+    public final void reportLoopCount(int iterations) {
```

Comment:
Well, technically, the branch profile needs runtime support to be useful, and that is provided by the runtime by knowing what `CompilerDirectives.injectBranchProbability(.)` is.

The difference from an implementation perspective is probably, that the `ConditionProfile` communicates information directly to the compiler, while the loop count is report to the more general runtime system which wants to decide earlier on when to compile stuff.